ethereum/p2p/src/test/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshStateTest.java

```diff
@@ -77,50 +96,348 @@
+    peer2.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.start(asList(peer1, peer2), TARGET);
+
+    verify(neighborFinder).findNeighbours(peer1, TARGET);
+    verify(neighborFinder).findNeighbours(peer2, TARGET);
+
+    verifyNoMoreInteractions(bondingAgent, neighborFinder);
+  }
+
+  @Test
+  public void shouldBondWithNewlyDiscoveredNodes() {
+    peer1.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.start(singletonList(peer1), TARGET);
+
+    verify(neighborFinder).findNeighbours(peer1, TARGET);
+    recursivePeerRefreshState.onNeighboursPacketReceived(
+        peer1, NeighborsPacketData.create(asList(peer2, peer3)));
+
+    verify(bondingAgent).performBonding(peer2);
+    verify(bondingAgent).performBonding(peer3);
+
+    verifyNoMoreInteractions(bondingAgent, neighborFinder);
+  }
+
+  @Test
+  public void shouldMoveToNeighboursRoundWhenBondingTimesOut() {
+    peer1.setStatus(PeerDiscoveryStatus.BONDED);
+    recursivePeerRefreshState.start(asList(peer1, peer2), TARGET);
+
+    verify(bondingAgent).performBonding(peer2);
+    timerUtil.runTimerHandlers();
+
+    verify(neighborFinder).findNeighbours(peer1, TARGET);
+
+    verifyNoMoreInteractions(bondingAgent, neighborFinder);
+  }
+
+  @Test
+  public void shouldMoveToBondingRoundWhenNeighboursRoundTimesOut() {
+    setupRelativePeerTree();
+
+    recursivePeerRefreshState.start(singletonList(peer_000), TARGET);
+
+    verify(bondingAgent).performBonding(peer_000);
+
+    peer_000.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.onBondingComplete(peer_000);
+
+    verify(neighborFinder).findNeighbours(peer_000, TARGET);
+
+    recursivePeerRefreshState.onNeighboursPacketReceived(peer_000, neighborsPacketData_000);
+
+    verify(bondingAgent).performBonding(peer_010);
+    verify(bondingAgent).performBonding(peer_011);
+    verify(bondingAgent).performBonding(peer_012);
+    verify(bondingAgent).performBonding(peer_013);
+
+    peer_010.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_011.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_012.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_013.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.onBondingComplete(peer_010); // Trigger the next round...
+
+    verify(neighborFinder, never()).findNeighbours(peer_010, TARGET);
+    verify(neighborFinder).findNeighbours(peer_011, TARGET);
+    verify(neighborFinder).findNeighbours(peer_012, TARGET);
+    verify(neighborFinder).findNeighbours(peer_013, TARGET);
+
+    timerUtil.runTimerHandlers();
+
+    verify(neighborFinder).findNeighbours(peer_010, TARGET);
+
+    timerUtil.runTimerHandlers();
+
+    recursivePeerRefreshState.onNeighboursPacketReceived(peer_011, neighborsPacketData_011);
+
+    verify(bondingAgent).performBonding(peer_120);
+    verify(bondingAgent).performBonding(peer_121);
+    verify(bondingAgent).performBonding(peer_122);
+    verify(bondingAgent).performBonding(peer_123);
```

Comment:
Can we write these with a simpler set of peers so we don't need so many different fields and the `peers.json` file?  There's nothing specific about the number of peers here - just need two peers to bond, then the first responds to neighbours request with peer3, while peer2 times out.  Check that we moved to bonding round and sent a bonding request to peer3.